test/over_react_test/utils/shadow_nested_component.dart

```diff
@@ -0,0 +1,43 @@
+// Copyright 2021 Workiva Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import 'dart:html';
+
+import 'package:over_react/over_react.dart';
+import 'package:over_react/react_dom.dart';
+
+part 'shadow_nested_component.over_react.g.dart';
+
+mixin ShadowNestedProps on UiProps {
+  String shadowRootHostTestId;
+  String shadowRootFirstChildTestId;
+  Ref shadowRootHostRef;
+}
+
+UiFactory<ShadowNestedProps> ShadowNested = uiFunction(
+  (props) {
+    final divRef = props.shadowRootHostRef ?? useRef<DivElement>();
+
+    useEffect(() {
+      if (divRef.current.shadowRoot == null) {
+        var shadowRootFirstChild = DivElement()..dataset['test-id'] = props.shadowRootFirstChildTestId ?? 'shadowRootFirstChild';
+        divRef.current.attachShadow({'mode':'open'}).append(shadowRootFirstChild);
+        render(Fragment()(props.children), shadowRootFirstChild);
+      }
+    }, [props]);
```

Comment:
I had a comment from earlier that I think got missed:

> Also, since we're using render the nested component will unmount itself, and no unmount call in the effect cleanup is necessary, but that might not be obvious to future maintainers. Could you add a comment?

Or, as opposed to going through the trouble of a comment, perhaps we should just unmount it normally?

Also, since this checks the ref, it only ever runs once, so this is effectively the same as using an effect with an empty dependencies list. We should just do that instead.
```suggestion
    useEffect(() {
      var shadowRootFirstChild = DivElement()..dataset['test-id'] = props.shadowRootFirstChildTestId ?? 'shadowRootFirstChild';
      divRef.current.attachShadow({'mode':'open'}).append(shadowRootFirstChild);
      react_dom.render(Fragment()(props.children), shadowRootFirstChild);
      return () => react_dom.unmountComponentAtNode(shadowRootFirstChild);
    }, []);
```